### ISSUE TITLE
Ruby Concurrent Jscall failure

### ISSUE THREAD

**Issue by adam429** [csg-tokyo/jscall#6]
When ruby code use Jscall, if the Ruby use multi-thread concurrent, the Jscall will be failed.

```require 'jscall'

# define a function in js
Jscall.exec <<-CODE
    async function test(x)
    {
        await new Promise(r => setTimeout(r, 1000));
        return x
    }
CODE


# concurrent call in js
Jscall.exec <<-CODE
    async function parallel() {
        let p = []
        for (var i=1;i<=20;i++) {
            p.push(test(i))
        }
        return await Promise.all(p)
    }
CODE
ret1 = Jscall.parallel()

puts ret1.to_s
# output [1,2,3,4,5 ... , 20], ok


# non-concurrent call in ruby
ret2 = (1..20).to_a.map do |x|
    Jscall.test(x)
end
puts ret2.to_s
# output [1,2,3,4,5 ... , 20], ok



# concurrent call in ruby
ret3 = (1..20).to_a.map do |x|
    Thread.new do
        Jscall.test(x)
    end
end.map(&:value)

puts ret3.to_s
# except [1,2,3,4,5 ... , 20], but error```

**Comment by chibash** [comment]
Yes, that's true.  The implementation of Jscall supports only single-threaded code in Ruby.

**Comment by adam429** [comment]
Is there any possibility that we could support multi-threaded code from Ruby call (As the example I showed)

My thought was like this:  in the send_command / send_reply data send from Ruby via PipeToJs, we add a session_id field to identify the session (The session_id is stay the same for multiple send_command / send_reply back-forth in one Ruby initiate Jscall.exec)

When the Ruby side Pipe receive the data in one thread (there could multiple threads initiate Jscall.exec with different session), the thread only fetch data in Pipe who has the same session_id.  (Like a Messaging Queue, receiver only subscribe the specific topic)

For the JS side Pipe, it can process all session_id data in the same main_loop.  But the process should be non-block via JS concurrency.

The Pipe between JS & Ruby is the shared resource.  So multiple session can use the Pipe together with mutex and topic subscribe.
